### DISCUSSION TITLE
How can a provider shared on multiple pages update only the current page?

### DISCUSSION THREAD

**Discussion by ericvalero** [rrousselGit/riverpod#1548] [Q&A]
I have an app with multiple pages and a filter that is shared on all of them. When the filter is changed it triggers an update on all the pages. 

As an example this are the providers that use the same filter (`filterProvider`):
```
final provider1 = FutureProvider<List<...>>((ref) async {
  final String filter = ref.watch(filterProvider);
  return ref.read(apiProvider).getFirstPageData(filter);
});

final provider2 = FutureProvider<List<...>>((ref) async {
  final String filter = ref.watch(filterProvider);
  return ref.read(apiProvider).getSecondPageData(filter);
});
```
This is the page that navigates to the second one:
```
class FirstPage extends StatelessWidget {
  const FirstPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final data = ref.watch(provider1);

    return Scaffold(
      appBar: AppBar(),
      body: Column(
        children: [
          const Filter(), // This widget uses filterProvider
          data.when(...),
          ElevatedButton(
            child: Text('Navigate'),
            onPressed: () => Navigator.of(context).push(MaterialPageRoute(
          builder: (context) => SecondPage(),
            )),
          ),
        ],
      ),
    );
  }
}
```
And the second page which is also using the same filter:
```
class SecondPage extends StatelessWidget {
  const SecondPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final data = ref.watch(provider2);

    return Scaffold(
      appBar: AppBar(),
      body: Column(
      children: [
        const Filter(), // This widget uses filterProvider
        data.when(...),
      ]),
    );
  }
}

```

When the filter is updated, how can I trigger a change only for the current page (top of the navigator stack)? 

I tried to use `pushReplacement` instead of `push` (as recommended [here](https://github.com/rrousselGit/riverpod/discussions/965)) and it works as expected but we lose the reference to the previous page, so we have to manually implement a back button which adds a lot of complexity.

Thanks.

**Comment by rrousselGit** [comment]
By definition, if a provider is shared between pages, updating it will update all the pages.

It sounds like you don't want a filter that is shared between pages.

**Reply by ericvalero** [reply]
Thanks for the response.
I need to use the same filter in multiple pages because the endpoints need to filter by the same value when navigate back or forward. I need to persist the filter value through pages.

What can be the solution for this situation? Do I need to create a filter provider for each page? Or use `ConsumerStatefulWidget` to save the state of the filter on each page?

**Comment by otto-dev** [comment]
Solution concept draft:

`provider1` and `provider2` have to become (for example) `StateNotifierProviders` with a method to receive the newest `filter` state, upon which they call the `apiProvider`. In your routing logic subscribe to `filterProvider` using `ref.listen` when you enter the route (instead of using `ref.watch(filterProvider)`), and unsubscribe when you leave the route. I suppose this must involve a `ref` provided by a `ConsumerState`. In the listener, push the changes that you receive from `filterProvider` to `provider1` or `provider2` (depending on the route).


This should work conceptually. If somebody has suggestions about best practices of how to subscribe/unsubscribe when entering/leaving a route, using common routing solutions, feel free to add them to this comment - because that's something I'm not sure about. I.e. how can the `ConsumerState` be notified about a route change, so that it can unsubscribe (and re-subscribe).
